### PR TITLE
Implement "xname_fe" and "xname_re" arguments in MixedLM.summary()

### PR DESCRIPTION
**Pull request by QuentinAndre**
The function `MixedLM.summary()` featured two arguments `xname_fe` and `xname_re` (respectively to change the display names of fixed effects and random effects), but the function did not use those values. As a consequence, those keywords did not have any impact.

This pull request implements those keywords. It tests that the length of the lists specified in `xname_fe` and `xname_re` match the number of fixed/random parameters, and insert them as index in the DataFrame used to summarize the results.

### CODE REVIEW

**Comment by josef-pkt** [comment]
Thanks,

Can you figure out a unit test that checks the names in the params table?

This looks safe enough to backport to 0.9.

**Comment by QuentinAndre** [comment]
Happy to do it. However, I don't have much experience with unit tests: could you summarize briefly what the test should do, or to what the output should be compared against?

**Comment by QuentinAndre** [comment]
Hi Josef,

The summary of MixedLM uses `summary2()` (rather than `summary()`), so I added tests in iolib/tests/test_summary2.py.

The only existing class was testing for LaTeX output: for cleanliness, I added three new tests to a separate class called `TestSummaryMixedLM`. This is my first time dabbling with unit tests, so hopefully what I did makes sense.

File: statsmodels/iolib/tests/test_summary2.py [hunk 1]
```diff
@@ -95,3 +95,59 @@ def test_OLSsummary(self):
...
+        x2 = [6, 2, 1, 0, 1, 4, 3, 8, 2, 1]
+        y = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+        df = pd.DataFrame({"y": y, "pid": pid, "x0": x0, "x1": x1, "x2": x2})
+        endog = df["y"].values
+        exog = df[["x0", "x1", "x2"]].values
+        groups = df["pid"]
+        summ = MixedLM(endog, exog, groups=groups).fit().summary(xname_fe=["Constant", "Age", "Weight"])
+
+        desired = ["Constant", "Age", "Weight"]
+        actual = summ.tables[1].index.values[0:3] # Second table is summary of params, first three values are FE
```
Comment (josef-pkt) [inline statsmodels/iolib/tests/test_summary2.py]: you could test all `summ.tables[1].index.values` each time
those should be deterministic

Comment (QuentinAndre) [reply statsmodels/iolib/tests/test_summary2.py]: Understood. I was not sure that the label `Group Var` for random effects would not change in the future, so I avoided including. Making the change now. Thanks for the feedback.

**Comment by josef-pkt** [comment]
Looks good overall.
unit tests for specific models should be with those models
i.e. add them to `statsmodels\regression\tests\test_lme.py`

The iolib test are mainly for the summary/summary2 implementation itself. but here we want to test MixedLM.

File: statsmodels/iolib/tests/test_summary2.py (hunk 1 above)
Comment (josef-pkt) [reply statsmodels/iolib/tests/test_summary2.py]: The advantage of the tests is that there are no future unintentional changes that possibly break current functionality. If those future changes are intentional, then the unit tests will need to be adjusted when those changes are made.

**Comment by QuentinAndre** [comment]
I have moved the unit tests to where they should be..

**Comment by josef-pkt** [comment]
let's wait for the Travis-CI unit tests to finish, then there are a few more changes to the tests.

(The PR will need squashing before merging.)

File: statsmodels/iolib/tests/test_summary2.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 from statsmodels.iolib.summary2 import summary_col
 from statsmodels.regression.linear_model import OLS, add_constant
-
+from statsmodels.regression.mixed_linear_model import MixedLM
```
Comment (josef-pkt) [inline statsmodels/iolib/tests/test_summary2.py]: remove this again

There should not be any changes to this module

Comment (QuentinAndre) [reply statsmodels/iolib/tests/test_summary2.py]: Crap, sorry I missed it. It's done.

**Comment by QuentinAndre** [comment]
TIL of setup_class, thanks. Much cleaner now. Sorry it took me so much time, first time with unit tests.

File: statsmodels/regression/mixed_linear_model.py [hunk 3]
```diff
@@ -2382,6 +2382,20 @@ def summary(self, yname=None, xname_fe=None, xname_re=None,
         info["Model:"] = "MixedLM"
         if yname is None:
             yname = self.model.endog_names
+        param_names = self.model.data.param_names[:]
+        n_fe_params = len(self.fe_params)
```
Comment (josef-pkt) [inline statsmodels/regression/mixed_linear_model.py]: minor naming convention change: change prefix "n" to "k"
n_fe_params -> k_fe_params
and similar for other cases

(the module use k_fe, but k_fe_params is more explicit here)

we try to keep "n" prefix for observation related counts, and "k" prefix for variable or

File: statsmodels/regression/mixed_linear_model.py [hunk 4]
```diff
@@ -2382,6 +2382,20 @@ def summary(self, yname=None, xname_fe=None, xname_re=None,
...
+        if xname_fe is None:
+            xname_fe = param_names[:n_fe_params]
+        if len(xname_fe) != n_fe_params:
+            raise ValueError("xname_fe should be a list of length %d" % n_fe_params)
+        if xname_re is None:
+            xname_re = param_names[n_fe_params:]
+        if len(xname_re) != n_re_params:
+            raise ValueError("xname_re should be a list of length %d" % n_re_params)
+        param_names[:n_fe_params] = xname_fe
+        param_names[n_fe_params:] = xname_re
```
Comment (josef-pkt) [inline statsmodels/regression/mixed_linear_model.py]: i think it would be better to change the logic here to only assign if xname_fe (by user) is not None, and don't change param_names if it is not needed
e.g
```
if xname_fe is not None:
  if len(xname_fe) != n_fe_params:
        raise ValueError("xname_fe should be a list of length %d" % n_fe_params)
  param_names[:n_fe_params] = xname_fe

if xname_re is not None:
...
```

Comment (QuentinAndre) [reply statsmodels/regression/mixed_linear_model.py]: All good points, and dully noted for the naming convention.

**Comment by josef-pkt** [comment]
This pull request **introduces 1 alert** when merging f00971e3acf1dba5f03b01f7faa02414b55d2478 into 5225ee53250ed87ae7e55fbf45ca96a4e22bf5b0 - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-25203dc045acdc9cf3581de43446e4c093770222)

**new alerts:**

* 1 for Unreachable code

---

*Comment posted by [lgtm.com](https://lgtm.com)*

File: statsmodels/regression/mixed_linear_model.py [hunk 5]
```diff
@@ -2382,6 +2382,18 @@ def summary(self, yname=None, xname_fe=None, xname_re=None,
...
+        k_fe_params = len(self.fe_params)
+        k_re_params = len(param_names) - len(self.fe_params)
+        if xname_fe is not None:
+            if len(xname_fe) != k_fe_params:
+                raise ValueError("xname_fe should be a list of length %d" % k_fe_params)
+            param_names[:k_fe_params] = xname_fe
+        if xname_re is not None:
+            if len(xname_re) != k_re_params:
+                raise ValueError("xname_re should be a list of length %d" % k_re_params)
+                param_names[k_fe_params:] = xname_re
```
Comment (josef-pkt) [inline statsmodels/regression/mixed_linear_model.py]: wrong indentation, test fails

**Comment by josef-pkt** [comment]
code looks good now, and unit test look like they are going to be green (I canceled some outdated appveyor job to avoid redundant test runs)

Can you squash all commits in an interactive rebase?
If not I will do it before merge, most likely later today.
(I'm going to be offline for the next few hours, but this looks ready to me.)

**Comment by QuentinAndre** [comment]
It's starting to be a crash-course in git and open-source development ;). I tried to squash all my commits into one, and rebased it, but I'm not sure I did things correctly.

**Comment by josef-pkt** [comment]
on github the squash and rebase looks good, it's now just a single commit.
I will check out the PR locally to make sure it is correctl

**Comment by QuentinAndre** [comment]
Thanks a lot, it was really helpful. I noticed a few parts of the codebase I could contribute to (namely the MediationModel class), but those are more significant changes. Will tackle it once I have more time on my hands!
